Develop dual fix customer

haiwentelink opened the pull request:
> !!!!!!!!!! Please delete the instructions below and replace with PR description
>
> If you have an issue number, please use a syntax of
> `Fixes #12345` and a brief change description
>
> If you do not have an issue number, please have a good description of
> the problem and the fix. Help the reviewer understand what to expect.
>
> Make sure you delete these instructions (to prove you have read them).
>
> !!!!!!!!!! Instructions end

damien0x0023 commented on config/telink/chip-module/Kconfig.defaults:
```diff
@@ -265,6 +265,11 @@ config COMPRESS_LZMA
     default n
 
+config CUSTOMER_MODE
+    bool "control the demo code"
```
Suggest to replace it with "Enable/Disable Customer Mode"

haiwentelink replied:
@damien0x0023  should not use enable , so change to control , or CI wil fail which will check enable words in descriptor.

damien0x0023 commented on examples/lighting-app/telink/src/ZclCallbacks.cpp:
```diff
@@ -33,6 +33,14 @@ using namespace chip::app::Clusters;
                                        uint8_t * value)
 {
+#if CONFIG_CUSTOMER_MODE
+    /* user mode , add the customer code here for cb*/
```
customer mode

haiwentelink replied:
will change

damien0x0023 commented on examples/platform/telink/common/include/AppTaskCommon.h:
```diff
@@ -43,9 +43,25 @@
+#elif CONFIG_SOC_RISCV_TELINK_TL321X
+#define ZB_NVS_PARTITION slot1_partition
+#define ZB_NVS_PARTITION_DEVICE FIXED_PARTITION_DEVICE(ZB_NVS_PARTITION)
+#define ZB_NVS_START_ADR FIXED_PARTITION_OFFSET(ZB_NVS_PARTITION)
```
If possible, we may move the line 49&50 after the condition clauses so that we can remove the line 54&55.

haiwentelink replied:
will change

damien0x0023 commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -143,6 +146,8 @@ void FactoryResetExtHandler(void)
 {
     // Erase the user parameters partition to reset mode settings
     flash_erase(flash_para_dev, USER_PARTITION_OFFSET, USER_PARTITION_SIZE);
+    // Need to erase zb nvs part in factory mode
```
"Erase ZigBee NVS data during factory reset"

haiwentelink replied:
will change

damien0x0023 commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -257,25 +273,28 @@ CHIP_ERROR AppTaskCommon::StartApp(void)
     if (user_para.val == USER_ZB_SW_VAL)
     {
         sBoot_zb = 1;
+        /* if switch from zb , need to get all the cluster info from zb */
```
If switching from ZigBee mode, get all existing clusters information.

haiwentelink replied:
will change

damien0x0023 commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -257,25 +273,28 @@ CHIP_ERROR AppTaskCommon::StartApp(void)
      * a given timeframe.
      */
     flash_read(flash_para_dev, USER_PARTITION_OFFSET, &user_para, sizeof(user_para));
+    /* Boot from Zigbee , need to clean the user parameters sector first and set a flag */
```
If booting from Zigbee mode, set a flag and clean deprecated parameter sectors.

haiwentelink replied:
will change

damien0x0023 commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -333,13 +352,16 @@ CHIP_ERROR AppTaskCommon::InitCommonParts(void)
 
     PrintFirmwareInfo();
+#if CONFIG_CUSTOMER_MODE
 
```
It would be better to add some description here to remind developers, e.g. "/* Leave empty condition clause for development on customers' boards */"

haiwentelink replied:
will change

damien0x0023 commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -823,6 +845,24 @@ void AppTaskCommon::ChipEventHandler(const ChipDeviceEvent * event, intptr_t /*
     case DeviceEventType::kCommissioningComplete: {
         uint8_t val = USER_MATTER_PAIR_VAL;
+#if CONFIG_CUSTOMER_MODE
+        /* need to add here to update the cluster information , only in the zb switch and touchlink is paired*/
```
Update all cluster information, only if switching from ZigBee mode and the touchlink is paired.

haiwentelink replied:
will change

damien0x0023 commented on examples/platform/telink/common/src/AppTaskCommon.cpp:
```diff
@@ -823,6 +845,24 @@ void AppTaskCommon::ChipEventHandler(const ChipDeviceEvent * event, intptr_t /*
+        if (user_para.val == USER_ZB_SW_VAL && user_para.on_net)
+        {
+            Protocols::InteractionModel::Status status;
+            /* Switch from the touch link, need to restore previous values */
```
If switching from the touch link, restore the previous light status values

haiwentelink replied:
will change

damien0x0023 reviewed (approved):
It's generally okay on code style.
Please add some content into this PR description.
Regarding the comments, you may refine them if have spare time.

haiwentelink committed:
riscv: telink: add dual mode for b92 and buteo and CI.

- clean the boot flag in init part .
- add the kDnsstimer to protect from init to dnss .
- add special proc for ikea.
- add the user-para proc in matter.
- add 2m flash for buteo ,and 4m for b92.
- update CI for b92 and tl3218x .
- fix the switch logic part .

Signed-off-by: Haiwen Xia <haiwen.xia@telink-semi.com>
